packages/components/src/autocomplete/index.js

```diff
@@ -231,39 +171,27 @@ export class Autocomplete extends Component {
 		this.node = node;
 	}
 
-	insertCompletion( range, replacement ) {
-		const container = document.createElement( 'div' );
-		container.innerHTML = renderToString( replacement );
-		while ( container.firstChild ) {
-			const child = container.firstChild;
-			container.removeChild( child );
-			range.insertNode( child );
-			range.setStartAfter( child );
-		}
-		range.deleteContents();
+	insertCompletion( replacement ) {
+		const { record, onChange } = this.props;
 
-		let inputEvent;
-		if ( undefined !== window.InputEvent ) {
-			inputEvent = new window.InputEvent( 'input', { bubbles: true, cancelable: false } );
-		} else {
-			// IE11 doesn't provide an InputEvent constructor.
-			inputEvent = document.createEvent( 'UIEvent' );
-			inputEvent.initEvent( 'input', true /* bubbles */, false /* cancelable */ );
-		}
-		range.commonAncestorContainer.closest( '[contenteditable=true]' ).dispatchEvent( inputEvent );
+		replacement = replacement.slice( 1 + this.state.query.length );
```

Comment:
No, it's to replace the prefix and the already typed characters to minimise updates. So a multi character trigger can still work, in fact, there's no longer such a thing, you can complete customise what to match and how to match it.

Reply:
Ah, OK. That helps.

There are a couple of assumptions here that may not hold for all completers.

This line relies on completions always finishing what is being typed (e.g., `+doto` being completed with `+dotorg`), but completions may be anything and aren't constrained to what the user typed. For example, a fruit completer could complete "!apple" with '🍎'. This means we need to replace the entire substring corresponding to the completer query rather than simply adding the part that isn't there.

It's also possible for completers to return `wp.element` declarations, so eventually this method needs to to account for that.

Reply:
That's a very good point, thank you!

Reply:
The apple example seems to be something that closer to a pattern like `` `test` ``... I wonder if we can somehow unify all of these little functionalities.

Reply:
> I wonder if we can somehow unify all of these little functionalities.

I like the direction you're going with the format configuration [here](https://github.com/WordPress/gutenberg/pull/7890#issuecomment-409623649). I'm not sure if the current completer interface will ultimately have a role here and need to sleep on it to give my mind some time to adjust. :)

Reply:
> It's also possible for completers to return `wp.element` declarations, so eventually this method needs to to account for that.

Since we're allowing matched tokens to be either selected from a list or typed in full, maybe we should reconsider allowing completions to be `wp.element` declarations. If we restrict to text completions, completions would be simpler, and there could be a single place to declare token formatting and a single execution path for applying that formatting.